server/src/main/java/org/elasticsearch/cluster/routing/IndexRouting.java

```diff
@@ -116,4 +205,158 @@ public void collectSearchShards(String routing, IntConsumer consumer) {
             }
         }
     }
+
+    private static class ExtractFromSource extends IndexRouting {
+        private final String indexName;
+        private final FilterPath[] include;
+
+        ExtractFromSource(int routingNumShards, int routingFactor, String indexName, List<String> routingPaths) {
+            super(routingNumShards, routingFactor);
+            this.indexName = indexName;
+            this.include = FilterPath.compile(Set.copyOf(routingPaths));
+        }
+
+        @Override
+        public int indexShard(
+            boolean idIsAutoGenerated,
+            String id,
+            @Nullable String routing,
+            XContentType sourceType,
+            BytesReference source
+        ) {
+            if (idIsAutoGenerated == false) {
+                throw new IllegalArgumentException(error("indexing with a specified id"));
+            }
```

Comment:
This seems like more of a timeseries db specific request validation, I see no reason to enforce auto-id otherwise? I wonder if it is a bit premature to add it here even, since the logic to build the `id` is not yet added. I think `IndexRequest.process` will need some kind of indication anyway that it is a timeseries db request and thus it may move there in the end? The only use of this variable is in this validation, so if we remove it here, the interface for `indexShard` will become simpler too.

Reply:
I can drop it for now but we need to do *something* about this before we remove the feature flag. Our hope was to disable auto-generated ID stuff for timeseries so we could deal with the all the plumbing of not generating the id at all at a later stage. But I can pull this out and we can keep it on the TODO list.